procwrap/procwrap.py

```diff
@@ -3,173 +3,208 @@
-            nsqd_http_address = 'http://%s' % nsqd_http_address
-
-        self.nsqd_http_address = nsqd_http_address
+# TODO make this configurable ?
+NSQ_MAX_CONTENT_LENGTH = 1024 * 1024 # 1 MB
 
-        self.process = None
+class ProcessWrapper(BaseScript):
+    DESC = "process wrapper"
+
+    def pre_run_init(self):
         self.write_to_nsq = lambda msg: None
+        self.process = None
+
+        if not self.args.enable_nsq:
+            return
 
-        if not self.disable_nsq:
-            self.init_nsq()
+        self.init_nsq()
 
     def init_nsq(self):
-        session = requests.Session()
-        # TODO use urljoin ?
-        response = None
-        try:
-            response = session.get("%s/ping" % self.nsqd_http_address)
-            if response.status_code != 200:
-                raise Exception("bad response %s" % response)
-        except:
-            logger.exception(
-                "if you wish to disable nsqd, use --disable-nsq"
-            )
-            sys.exit(-1)
-
-        # TODO nsq is a failure point for our logs. what do we do if nsq is down ?
-
-        url = "%s/pub" % self.nsqd_http_address
-        topic = self.nsq_log_topic
-        # TODO what if this write to nsq fails ?
-        self.write_to_nsq = lambda msg: session.post(url, data=msg, params={"topic": topic})
-
-    def start(self):
         """
-        starts the command and blocks till it finishes.
-        once finished, returns the returncode.
+        if --enable-nsq is provided, initialize the nsq
         """
 
-        if self.pid_file is not None:
-            d = os.path.dirname(self.pid_file)
-            if not os.path.exists(d):
-                os.makedirs(d)
+        addr = self.args.nsqd_http_address
+        addr = addr.rstrip('/')
+        if not addr.startswith('http://'):
+            addr = 'http://%s' % addr
+
+        self.args.nsqd_http_address = addr
+
+        response = requests.get('%s/ping' % self.args.nsqd_http_address)
+        if response.status_code != 200:
+            raise Exception("bad response %s" % response)
 
-        # NOTE only capturing stderr. stdout still goes to tty.
+        # TODO make queuesize configurable
+        queue = Queue.Queue(maxsize=1000)
+        # block until you can write it to the queue
+        self.write_to_nsq = lambda msg: queue.put(msg, block=True, timeout=None)
 
+        self.queue = queue
+        self.nsq_publish_thread = threading.Thread(target=self._publish_to_nsq)
+        self.nsq_publish_thread.daemon = True
+        self.keeprunning = threading.Event()
+        self.keeprunning.set()
+
+        self.nsq_publish_thread.start()
+
+    def _publish_to_nsq(self):
+        url = '%s/mpub' % self.args.nsqd_http_address
+        topic = self.args.nsq_log_topic
+        self.log.info("pushing logs to nsq", url=url, topic=topic)
+
+        # TODO make this configurable
+        interval = 5 # every five seconds or nsq max content length
```

Comment:
@algrebe Yeah, will be good to have this configurable via command-line args. Also, default should be 500ms. Frequent enough and unlikely to cause significant resource consumption.